FirebaseSegmentation/Sources/SEGDatabaseManager.m

```diff
@@ -0,0 +1,419 @@
+  dispatch_async(_databaseOperationQueue, ^{
+    SEGDatabaseManager *strongSelf = weakSelf;
+    if (!strongSelf) {
+      completionHandler(NO, @{@"ErrorDescription" : @"Internal database error"});
+    }
+    NSString *dbPath = [SEGDatabaseManager pathForSegmentationDatabase];
+    // TODO(dmandar) log.
+    NSLog(@"Loading segmentation database at path %@", dbPath);
+    const char *databasePath = dbPath.UTF8String;
+    // Create or open database path.
+    if (!SegmentationCreateFilePathIfNotExist(dbPath)) {
+      completionHandler(NO, @{@"ErrorDescription" : @"Could not create database file at path"});
+    }
+    int flags = SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FILEPROTECTION_COMPLETE |
+                SQLITE_OPEN_FULLMUTEX;
+    if (sqlite3_open_v2(databasePath, &strongSelf->_database, flags, NULL) == SQLITE_OK) {
+      // Create table if does not exist already.
+      if ([strongSelf createTableSchema]) {
+        // DB file created or already exists.
+        // Exclude the app data used from iCloud backup.
+        SegmentationAddSkipBackupAttributeToItemAtPath(dbPath);
+
+        // Read the database into memory.
+        NSDictionary<NSString *, NSString *> *associations = [self loadMainTable];
+        completionHandler(YES, associations);
+
+      } else {
+        // Remove database before fail.
+        [strongSelf removeDatabase:dbPath];
+        FIRLogError(kFIRLoggerSegmentation, @"I-SEG000010", @"Failed to create table.");
+        // Create a new database if existing database file is corrupted.
+        if (!SegmentationCreateFilePathIfNotExist(dbPath)) {
+          completionHandler(NO,
+                            @{@"ErrorDescription" : @"Could not recreate database file at path"});
+        }
+        if (sqlite3_open_v2(databasePath, &strongSelf->_database, flags, NULL) == SQLITE_OK) {
+          if (![strongSelf createTableSchema]) {
+            // Remove database before fail.
+            [strongSelf removeDatabase:dbPath];
+            // If it failed again, there's nothing we can do here.
+            FIRLogError(kFIRLoggerSegmentation, @"I-SEG000010", @"Failed to create table.");
+          } else {
+            // Exclude the app data used from iCloud backup.
+            SegmentationAddSkipBackupAttributeToItemAtPath(dbPath);
+          }
+        } else {
+          [strongSelf logDatabaseError];
+          completionHandler(NO, @{@"ErrorDescription" : @"Could not create database."});
+        }
+      }
+    } else {
+      [strongSelf logDatabaseError];
+      completionHandler(NO, @{@"ErrorDescription" : @"Error creating database."});
+    }
+  });
+}
+
+- (void)removeDatabase:(NSString *)path completion:(SEGRequestCompletion)completionHandler {
+  __weak SEGDatabaseManager *weakSelf = self;
+  dispatch_async(_databaseOperationQueue, ^{
+    SEGDatabaseManager *strongSelf = weakSelf;
+    if (!strongSelf) {
+      return;
+    }
+    [strongSelf removeDatabase:path];
+  });
```

Comment:
`completionHandler` should be called before the block exits.

Reply:
done